managed-pointer/cpp-20/D0567R1.md

```diff
@@ -5,222 +5,384 @@
 
-The first example shows how this would look using the explicit interface for synchronisation operations (figure 3) and the second shows how this would look using the implicit interface for synchronisation operations (Figure 4).
+If a user wishes to explicitly allocate or deallocate memory for a `managed_ptr` instance on a particular execution context, this can be done by calling the allocation operations `allocate`, `then_allocate`, `deallocate` and `then_deallocate`. Each of which is an asycnhronous operation which performs an allocation or dallocation on a particular execution context and returns a future to the complletion of the operation.
+
+If any memory allocation or deallocation operations fail then an exception is thrown and stored in the future object that is returned by the operations which triggered the allocation operation.
+
+#### 3.3.5. Synchronisation
+
+For an **execution context** to be the accessible **execution context**, if it is not already, a synchronisation operation is required. A synchronisation operation is an implementation defined asynchronous st of commands which moves the data from the currently accessible memory region to another memory region which returns a future object that can be used to wait for the operation to complete. From the point at which a synchronisation operation is triggered the currently accessible memory region is no longer accessible. Once the synchronisation point is complete (i.e. the future returned from the synchronisation operation is complete) the memory region the data is being synchronised to is now the accessible memory region.
+
+Synchronisation operations are course grained synchronisation in that they synchronise the entire managed memory allocation of a `managed_ptr`.
+
+> This limited may be relaxed in a future version of the paper when a `managed_ptr` can be concurrently accessed in more than one memory region. In this case there sychronisation could be performed at smaller granularities using mapping operations or atomic operations. This is further described in section 5.1.
+
+There are three sycnhronisation operations; `get`, `put` and `sync`, each taking one or more `managed_ptr` parameters. These sycnhronisation operations perform the following sycnhronisation for each `managed_ptr` parameter:
+
+* The `get` operation synchronises the data at the memory region at a particular *execution context* to the memory region at the *execution context* associated with the `managed_ptr`.
+* The `put` operation synchronises the data at the memory region at the *execution context* associated with the `managed_ptr` to the memory region at a particular *execution context*.
+* The `sync` operation synchronises the data at the memory region at one particular *execution context* to the memory region at another particular *execution context*.
```

Comment:
The reason for maintaining the `put` and `get` functions as well as `sync` is that all of these functions take a variadic pack of `managed_ptr` as a parameter and the host executor could potentially be different for each. Though it's most common that they would be the same.

We could perhaps simplify the interface to just the `sync` function:

```cpp
  template <typename SynchronizationChannel, typename SrcExecutor = managed_ptr<dataT>::host_executor_type,
            typename DestExecutor, typename dataT>
  future_type_t<void>
  sync(SynchronizationChannel = __default_synchronization_channel__{},
       SrcExecutor, DestExecutor, managed_ptr<dataT> managedPtr);
```

Though this would mean that only one managed_ptr could be passed to `sync` at a time and would perhaps  reduce the clarity of the semantics.